ratelimits/limiter.go

```diff
@@ -238,90 +164,212 @@ func (l *Limiter) Spend(ctx context.Context, txn Transaction) (*Decision, error)
+	ctx = context.WithoutCancel(ctx)
+	tats, err := l.source.BatchGet(ctx, bucketKeys)
+	if err != nil {
+		return nil, err
+	}
+
+	start := l.clk.Now()
+	batchDecision := newBatchDecision()
+	newTATs := make(map[string]time.Time)
+
+	for _, txn := range batch {
+		tat, exists := tats[txn.bucketKey]
+		if !exists {
+			// First request from this client.
+			tat = l.clk.Now()
+		}
+
+		d := maybeSpend(l.clk, txn.limit, tat, txn.cost)
+
+		if txn.limit.isOverride {
+			utilization := float64(txn.limit.Burst-d.Remaining) / float64(txn.limit.Burst)
+			l.overrideUsageGauge.WithLabelValues(txn.limit.name.String(), txn.bucketKey).Set(utilization)
+		}
+
+		if d.Allowed && (tat != d.newTAT) && txn.spend {
+			// New bucket state should be persisted.
+			newTATs[txn.bucketKey] = d.newTAT
+		}
+
+		if txn.spendOnly() {
+			d = allowedDecision
+		}
+		batchDecision.merge(d)
+	}
+
+	if batchDecision.Allowed {
+		err = l.source.BatchSet(ctx, newTATs)
+		if err != nil {
+			return nil, err
+		}
+		l.spendLatency.WithLabelValues("batch", Allowed).Observe(l.clk.Since(start).Seconds())
+	} else {
+		l.spendLatency.WithLabelValues("batch", Denied).Observe(l.clk.Since(start).Seconds())
+	}
+	return batchDecision.Decision, nil
+}
+
 // Refund attempts to refund all of the cost to the capacity of the specified
 // bucket. The returned *Decision indicates whether the refund was successful
 // and represents the current state of the bucket. The new bucket state is
 // persisted to the underlying datastore, if applicable, before returning. If no
-// bucket exists it will NOT be created.
+// bucket exists it will NOT be created. Spend-only Transactions are assumed to
+// be refundable. Check-only Transactions are never refunded.
 //
 // Note: The amount refunded cannot cause the bucket to exceed its maximum
 // capacity. Partial refunds are allowed and are considered successful. For
 // instance, if a bucket has a maximum capacity of 10 and currently has 5
 // requests remaining, a refund request of 7 will result in the bucket reaching
 // its maximum capacity of 10, not 12.
 func (l *Limiter) Refund(ctx context.Context, txn Transaction) (*Decision, error) {
-	if txn.cost <= 0 {
-		return nil, ErrInvalidCost
-	}
-
-	limit, err := l.getLimit(txn.limit, txn.bucketKey)
-	if err != nil {
-		if errors.Is(err, errLimitDisabled) {
-			return disabledLimitDecision, nil
-		}
-		return nil, err
+	if txn.allowOnly() {
+		return allowedDecision, nil
 	}
-
 	// Remove cancellation from the request context so that transactions are not
 	// interrupted by a client disconnect.
 	ctx = context.WithoutCancel(ctx)
 	tat, err := l.source.Get(ctx, txn.bucketKey)
 	if err != nil {
 		return nil, err
 	}
-	d := maybeRefund(l.clk, limit, tat, txn.cost)
-	if !d.Allowed {
-		// The bucket is already at maximum capacity.
-		return d, nil
+	d := maybeRefund(l.clk, txn.limit, tat, txn.cost)
+	if tat == d.newTAT || txn.checkOnly() {
+		return maybeRefund(l.clk, txn.limit, tat, 0), nil
```

Comment:
This stanza was confusing to me at first. In the function comment it says "Check-only Transactions are never refunded." But here it shows that a check-only transaction calls `maybeRefund`; a contradiction!

But then I see that `maybeRefund` is called with a cost of 0, and you've tweaked `maybeRefund` so it can handle a cost of 0. But what does `maybeRefund` do with a cost of 0?

If I can guess: `maybeRefund` with a cost of zero builds a Decision that shows the current state of the bucket based on the limit and the TAT fetched from storage? Is there a way to pull that out of `maybeRefund`?